core/modules/qdisp/QueryRequest.h

```diff
@@ -117,36 +115,32 @@ class QueryRequest : public XrdSsiRequest {
     virtual void ProcessResponseData(char *buff, int blen, bool last);
 
     void cancel();
+    bool getCancelled();
 
-    void cleanup();
+    void cleanup(); ///< Must be called when this object is no longer needed.
 
 private:
-    bool cancelled();
-
+    void _callMarkComplete(bool success);
     bool _importStream();
     bool _importError(std::string const& msg, int code);
     void _errorFinish(bool shouldCancel=false);
     void _finish();
-    void _registerSelfDestruct();
 
     XrdSsiSession* _session;
 
-    std::string _payload; ///< Request buffer
-    std::shared_ptr<ResponseRequester> _requester; ///< Response requester
+    std::shared_ptr<JobQuery> _jobQuery; ///< Job information.
+    JobDescription& _jobDesc; ///< Convenience reference to JobDescription inside _jobQuery.
 
-    /// To be called when the request completes
-    std::shared_ptr<util::UnaryCallable<void, bool> > _finishFunc;
-    /// To be called to retry a failed request
-    std::shared_ptr<util::VoidCallable<void> > _retryFunc;
-    /// Reference to an updatable Status
-    JobStatus& _status;
+    util::Flag<bool> _retried; ///< Protect against multiple retries of _jobQuery from a single QueryRequest.
+    util::Flag<bool> _calledMarkComplete; ///< Protect against multiple calls to MarkCompleteFunc
+                                          /// from a single QueryRequest.
 
     std::mutex _finishStatusMutex;
-    enum FinishStatus { ACTIVE, FINISHED, CANCELLED, ERROR } _finishStatus;
+    enum FinishStatus { ACTIVE, FINISHED, CANCELLED, ERROR } _finishStatus; // _finishStatusMutex
```

Comment:
I delete that. I replaced CANCELLED with the  _cancelled flag. Changing the state from ACTIVE to CANCELLED resulted in  logic in _errorFinish being skipped and it complicated other things as well.